File: server/java/src/main/java/app/inventory/Inventory.java
package app.inventory;

import com.stripe.Stripe;
import com.stripe.exception.StripeException;
import com.stripe.model.Product;
import com.stripe.model.Sku;
import com.stripe.model.ProductCollection;
import com.stripe.model.SkuCollection;
import sun.reflect.generics.reflectiveObjects.NotImplementedException;

import java.util.*;


public class Inventory {

    public static Long calculatePaymentAmount(ArrayList<Map> items) throws StripeException {
        // Get the Stripe Key
        Stripe.apiKey = System.getenv("STRIPE_SECRET_KEY");

        Long total = 0L;
        Integer quantity = 0;

        for (Map product: items) {
            Sku sku = Sku.retrieve(product.get("parent").toString());
            quantity = ((Double) product.get("quantity")).intValue();
            total += sku.getPrice() * quantity;
        }

        return total;
    }

    public static Long getShippingCost(String id) {
        if (id.equals("free")) {
            return 0L;
        } else if (id.equals("express")) {
            return 500L;
        } else {
            return 0L;
        }
    }

    public static SkuCollection listSkus(String productId) throws StripeException{
        // Get the Stripe Key
        Stripe.apiKey = System.getenv("STRIPE_SECRET_KEY");

        Map<String, Object> skuParams = new HashMap<String, Object>();
        skuParams.put("product", productId);
        skuParams.put("limit", "3");

        SkuCollection collection = Sku.list(skuParams);

        return collection;
    }

    public static ProductCollection listProducts() throws StripeException{
        // Get the Stripe Key
        Stripe.apiKey = System.getenv("STRIPE_SECRET_KEY");

        Map<String, Object> productParams = new HashMap<String, Object>();
        productParams.put("limit", "3");

        ProductCollection collection = Product.list(productParams);

        return collection;
    }

    public static Product retrieveProduct(String id) throws StripeException {
        // Get the Stripe Key
        Stripe.apiKey = System.getenv("STRIPE_SECRET_KEY");

        return Product.retrieve(id);
    }
}

File: server/go/inventory/inventory.go
package inventory

import (
	"fmt"

	"github.com/stripe/stripe-go"
	"github.com/stripe/stripe-go/product"
	"github.com/stripe/stripe-go/sku"
)

type Item struct {
	Parent   string
	Quantity int
}

type Listing struct {
	Data interface{} `json:"data"`
}

func ListProducts() (*Listing, error) {
	products := []*stripe.Product{}

	params := &stripe.ProductListParams{}
	params.Filters.AddFilter("limit", "", "3")
	i := product.List(params)
	for i.Next() {
		products = append(products, i.Product())
	}

	err := i.Err()
	if err != nil {
		return nil, fmt.Errorf("inventory: error listing products: %v", err)
	}

	return &Listing{products}, nil
}

func ListSKUs(productID string) (*Listing, error) {
	skus := []*stripe.SKU{}

	params := &stripe.SKUListParams{}
	params.Filters.AddFilter("limit", "", "1")
	i := sku.List(params)
	for i.Next() {
		skus = append(skus, i.SKU())
	}

	err := i.Err()
	if err != nil {
		return nil, fmt.Errorf("inventory: error listing SKUs: %v", err)
	}

	return &Listing{skus}, nil

}

func CalculatePaymentAmount(items []Item) (int, error) {
	return 0, nil
}
